src/rpc/budget.cpp

```diff
@@ -48,14 +48,53 @@ void budgetToJSON(CBudgetProposal* pbudgetProposal, UniValue& bObj)
     bObj.push_back(Pair("fValid", pbudgetProposal->fValid));
 }
 
-UniValue preparebudget(const UniValue& params, bool fHelp)
+void checkBudgetInputs(const UniValue& params, std::string &strProposalName, std::string &strURL,
+                       int &nPaymentCount, int &nBlockStart, CBitcoinAddress &address, CAmount &nAmount)
 {
     int nBlockMin = 0;
     CBlockIndex* pindexPrev = chainActive.Tip();
 
+    if (strProposalName.size() > 20)
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid proposal name, limit of 20 characters.");
+
+    strURL = SanitizeString(params[1].get_str());
+    std::string strErr;
+    if (!validateURL(strURL, strErr))
+        throw JSONRPCError(RPC_INVALID_PARAMETER, strErr);
+
+    nPaymentCount = params[2].get_int();
+    if (nPaymentCount < 1)
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid payment count, must be more than zero.");
+
+    // Start must be in the next budget cycle
+    if (pindexPrev != NULL) nBlockMin = pindexPrev->nHeight - pindexPrev->nHeight % Params().GetBudgetCycleBlocks() + Params().GetBudgetCycleBlocks();
+
+    nBlockStart = params[3].get_int();
+    if (nBlockStart % Params().GetBudgetCycleBlocks() != 0) {
+        int nNext = pindexPrev->nHeight - pindexPrev->nHeight % Params().GetBudgetCycleBlocks() + Params().GetBudgetCycleBlocks();
+        throw JSONRPCError(RPC_INVALID_PARAMETER, strprintf("Invalid block start - must be a budget cycle block. Next valid block: %d", nNext));
+    }
+
+    int nBlockEnd = nBlockStart + (Params().GetBudgetCycleBlocks() * nPaymentCount); // End must be AFTER current cycle
+
+    if (nBlockStart < nBlockMin)
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid block start, must be more than current height.");
+
+    if (nBlockEnd < pindexPrev->nHeight)
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid ending block, starting block + (payment_cycle*payments) must be more than current height.");
```

Comment:
Isn't this unreachable code?
If we are here it means 
```
nBlockEnd = nBlockStart + (...) > nBlockStart  >= nBlockMin >= pindexPrev->nHeight
```
thus `nBlockEnd > pindexPrev->nHeight`

Reply:
Correct; these are the things I saw when I got deeper into this and thus #965 turned up.

Reply:
Oh, right. Guess I'm getting confused with the order of these PRs.
Shouldn't 965 be rebased on top of this one then?

Reply:
Trust me i keep getting confused with them all too.   I can merge that one in here so this can be the final picture, but have the 3 separate commits to reflect the distinctions until they start going into mainline

Reply:
That would be fine by me.

Other solutions could be
1) rebase 965 on top of this one (and have merge order 964 --> 950 --> 965)
2) rebase this one on top of 965 (and merge like 964 --> 965 --> 950)

Reply:
#2 is pretty much the  likely scenario; since that -should- be the merge order.